### DISCUSSION TITLE
MAUI

### DISCUSSION THREAD

**Discussion by tig** [tui-cs/Terminal.Gui#1738] [Ideas]
https://devblogs.microsoft.com/dotnet/introducing-dotnet-maui-one-codebase-many-platforms/

I've been very lightly thinking about how terminal UI apps could be built with MAUI. 

I have an itch to dive deeper. Before I do, have any of y'all also been thinking about this?

If I had to sum up my vision, it'd be

"Build an app using .NET MAUI and easily have it run as a native terminal UI app on any platform Terminal.Gui is supported on."

Would people actually be interested in this, or is just a solution looking for a problem?

Y'all are probably aware that PowerShell's built-in terminal UI components (namely `Out-ConsoleGridView') are already built using Terminal.Gui. The traditional GUI components of PowerShell (e.g. `Out-GridView`) are built using Avalon.  The PowerShell team has stated they are abandoning these and will use MAUI in the future. This sounds like an ideal place to start Terminal.Gui work on MAUI as it's a very real customer.

**Comment by tig** [comment]
See here for more on Powershell: https://github.com/PowerShell/GraphicalTools/issues/101

**Comment by tznind** [comment]
This definitely sounds interesting.  What would be required to add this support?  Would it be something like this (but for the MAUI UI API layer):

https://github.com/jsuarezruiz/xamarin-forms-gui.cs

If I've understood correctly reading the docs on MAUI that you linked.  Is the goal to write xaml and have it run either a OS app (e.g. in windows) or in terminal gui? 

I think its a cool idea if it can be pulled off.  I currently have an app where there are 2 parallel implementations of many UIs (one WinForms and one Terminal.Gui) and have briefly looked into how they might be merged.  At the time I was thinking about writing a terminal gui implementation of [Eto](https://github.com/picoe/Eto) but didn't get much beyond tinkering.

**Reply by BDisp** [reply]
I have a pull request there a long time ago :-) It is more or less similar. The `Xamarin` will be supersede with the `MAUI`. But much of integration will be as the `WebConsole` is being managed.

**Comment by beedell-roke** [comment]
Are you suggesting that if this proposition is accepted, one codebase might be possible to render via the terminal and outside of it, graphically?

I ask because I'm searching for the easiest way to make a graphical interface to my PowerShell scripts that is cross-platform and supports the terminal, too. If this might provide that, I'll leave myself subscribed, since this is a long-term goal anyway.

**Reply by BDisp** [reply]
> Are you suggesting that if this proposition is accepted, one codebase might be possible to render via the terminal and outside of it, graphically?
> 

No, the changes in the `Terminal.Gui` was to overpass the run loop. Now it's possible to interact with any driver that is passed to the `Application.Init` that doesn't support `SyncronizationContext`. The `WebConsole` (https://github.com/Blazor-Console/HACC) uses the `Blazor.Canvas` to render to the screen. So, for `MAUI` probably must be implemented more or less like the https://github.com/jsuarezruiz/xamarin-forms-gui.cs.

> I ask because I'm searching the easiest way to make a graphical interface to my PowerShell scripts that is cross-platform and supports the terminal, too. If this might provide that, I'll leave myself subscribed since this is a long-term goal anyway.

I would like to see that :-)

**Reply by tznind** [reply]
I think this would be an incredible feature to be able to build with MAUI and compile for terminal as well as desktop/mobile.

Does anyone have an idea on the roadmap features we would need to support this?  I don't know nearly enough about MAUI to understand whats needed but is it something like?

1. Implement required core features in gui.cs (e.g. Layouts?)
2. Fork https://github.com/jsuarezruiz/xamarin-forms-gui.cs and update for MAUI?
3. Add all the controls that are missing from `xamarin-forms` (tree view etc)

**Reply by BDisp** [reply]
> I think this would be an incredible feature to be able to build with MAUI and compile for terminal as well as desktop/mobile.
> 
I also would like to see all this integrated with the `Terminal.Gui`.

> Does anyone have an idea on the roadmap features we would need to support this? I don't know nearly enough about MAUI to understand whats needed but is it something like?
> 
>     1. Implement required core features in gui.cs (e.g. Layouts?)
> 
I think is better implement all the require integration on the `gui.cs` to be more easily create new `MAUI` app with the `Terminal.Gui` Api.

>     2. Fork https://github.com/jsuarezruiz/xamarin-forms-gui.cs and update for MAUI?
> 
I already tried that, by replacing all the` Xamarim.Forms` Api to the `MAUI` Api, but I get a bunch of errors that is not very easy to fix. I think that will be later issued a migration tool, but I don't know when.

>     3. Add all the controls that are missing from `xamarin-forms` (tree view etc)

I hope that could be implemented a script that could map all the `Terminal.Gui` core to the `MAUI`.

`MAUI` have a template that can also create a desktop/mobile from a `Blazor` app. For the `WebConsole` (https://github.com/Blazor-Console/HACC) that may be possible.

**Comment by johnmbaughman** [comment]
This is conversation is bit old, but I figured I'd start here. I started a Roslyn based project to parse and create code using a XAML like format called XTUI. It's quite obviously in the infancy stages and needs a lot of work. I'm trying to keep things using concepts related to XAML and elements are all direct TGUI view objects. 

I've been tracking all progress in the README file incorporating tests and benchmarks as well as beginning to port some example projects to use XTUI. This has a long way to go, but I'm open to suggestions and guidance as to how to proceed. 

[Terminal.Gui.Xtui](https://github.com/johnmbaughman/Terminal.Gui.Xtui)

-John

**Reply by tig** [reply]
> This is conversation is bit old, but I figured I'd start here. I started a Roslyn based project to parse and create code using a XAML like format called XTUI. It's quite obviously in the infancy stages and needs a lot of work. I'm trying to keep things using concepts related to XAML and elements are all direct TGUI view objects.
> 
> I've been tracking all progress in the README file incorporating tests and benchmarks as well as beginning to port some example projects to use XTUI. This has a long way to go, but I'm open to suggestions and guidance as to how to proceed.
> 
> [Terminal.Gui.Xtui](https://github.com/johnmbaughman/Terminal.Gui.Xtui)
> 
> -John

Wicked cool!

**Reply by johnmbaughman** [reply]
Just a little update: keeping up with TGUI is quite the fun task! 👯 Had to take a few weeks off the project for personal stuff and fell so far behind... Time to catch up!
